infrastructure/src/main/java/org/corfudb/infrastructure/orchestrator/AddNodeWorkflow.java

```diff
@@ -0,0 +1,205 @@
+
+        final long chunkSize = 2500;
+
+        long trimMark = runtime.getAddressSpaceView().getTrimMark();
+        if (trimMark > segment.getEnd()) {
+            log.info("stateTransfer: Nothing to transfer, trimMark {} greater than end of segment {}",
+                    trimMark, segment.getEnd());
+            return;
+        }
+
+        for (long chunkStart = segment.getStart(); chunkStart < segment.getEnd()
+                ; chunkStart = chunkStart + chunkSize) {
+            long chunkEnd = Math.min((chunkStart + chunkSize - 1), segment.getEnd() - 1);
+
+            Map<Long, ILogData> dataMap = runtime.getAddressSpaceView()
+                    .cacheFetch(ContiguousSet.create(
+                            Range.closed(chunkStart, chunkEnd),
+                            DiscreteDomain.longs()));
+
+            List<LogData> entries = new ArrayList<>();
+            for (long x = chunkStart; x <= chunkEnd; x++) {
+                if (dataMap.get(x) == null) {
+                    log.error("Missing address {} in range {}-{}", x, chunkStart, chunkEnd);
+                    throw new IllegalStateException("Missing address");
+                }
+                entries.add((LogData) dataMap.get(x));
+            }
+
+            // Write segment chunk to the new logunit
+            boolean transferSuccess = runtime
+                    .getRouter(endpoint)
+                    .getClient(LogUnitClient.class)
+                    .writeRange(entries).get();
+
+            if (!transferSuccess) {
+                log.error("stateTransfer: Failed to transfer {}-{} to {}", chunkSize,
+                        chunkEnd, endpoint);
+                throw new IllegalStateException("Failed to transfer!");
+            }
+
+            log.info("stateTransfer: Transferred address chunk [{}, {}]",
+                    chunkStart, chunkEnd);
+        }
+    }
+
+
+    /**
+     * Copies the split segment to the new node, if it
+     * is the new node also participates as a logging unit.
+     */
+    class StateTransfer implements Action {
+        @Override
+        public String getName() {
+            return "StateTransfer";
+        }
+
+        @Override
+        public void execute(@Nonnull CorfuRuntime runtime) throws Exception {
+            // Transfer the replicated segment to the new node
+            stateTransfer(request.getEndpoint(), runtime, newLayout.getSegment(0));
+        }
+    }
+
+    /**
+     * Merges the fragmented segment if the AddNodeToLayout action caused any
+     * segments to split
+     */
+    class MergeSegments implements Action {
+        @Override
+        public String getName() {
+            return "MergeSegments";
+        }
+
+        @Override
+        public void execute(@Nonnull CorfuRuntime runtime) throws Exception {
+            // Transfer the replicated segment to the new node
+            // TODO(Maithem) skip this step if there are no disjoint segments
```

Comment:
![INFO](https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO') Complete the task associated to this TODO comment. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1135)